### PR TITLE
docs / oidc / Authorization code example with dotnet core

### PR DESCRIPTION
**Pull request by charleslevesque**
# Overview
This adds a new integration example with Azure AD OpenID Connect authentication. It logins the user for every page and makes a GraphQL request to WorkflowGen's API to get some of the user's profile information.

There's a lot of files because the `dotnet new` tool scaffolds a new project with bootstrap and jquery. The following files are the ones that should be reviewed:

* Pages/*
* appsettings.*
* AuthPropertiesTokenCache.cs
* Startup.cs
* oidc.csproj
* Utilities.Tokens.cs

## Used Libraries
* [GraphQL.Client](https://github.com/graphql-dotnet/graphql-client)

### CODE REVIEW

**Comment by charleslevesque** [comment]
Documentation for this example is missing but the code can be reviewed.

**Comment by jiananhsu** [comment]
@charleslevesque 

Can you provide an overview description of the architecture of your solution, the main modules or files pertaining to the OIDC auth management and the GraphQL call?

**Comment by charleslevesque** [comment]
# Authentication
This example uses the built-in solution from Microsoft to handle authentication. It has an OpenID Connect authentication provider. It is configured in Setup.cs. For more information about how identity works in aspnet core, see the [following link](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/identity?view=aspnetcore-2.2&tabs=visual-studio). To store the tokens, there is a built-in mechanism with the class TokenCache which uses a store (in this example, I use Cookies to keep the app stateless). I created a class (AuthPropertiesTokenCache) which derives from TokenCache to make manipulations easier. You can follow the link in the doc in the file to see where I picked up this example.

# GraphQL
There is only 1 call and it is on each GET to the WorkflowGenProfile.cshtml page (see the model WorkflowGenProfile.cshtml.cs). It uses the library GraphQL.Client which can be used dynamically or with types like in the example. I use the built-in dependency injection system to get a graphql client with pre-configured authorization header (see Startup.cs) to get an instance at each request.

**Comment by charleslevesque** [comment]
@pnicoll Could you review the README.md file of this PR? Here's the link:

https://github.com/advantys/workflowgen-templates/pull/6/files#diff-b73761b27e57651d166ee62bdac5aa6a

You can click on `View File` to get the formatted Markdown. Just add line comments for your comments, I will edit. You can add a line comment by clicking the `+` button appearing at the beginning of a line in the review page.

File: integration/oidc/authorization-code/README.md [hunk 1]
```diff
@@ -0,0 +1,197 @@
+# OpenID Connect Example with Server Side Web App
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: # OpenID Connect Example with Server-Side Web App

File: integration/oidc/authorization-code/README.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+## Overview
+This example application uses the OpenID Connect (OIDC) protocol to authenticate
+with Azure Active Directory and get an access token to request WorkflowGen's
+GraphQL API.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: This sample application uses the OpenID Connect (OIDC) protocol to authenticate with Azure Active Directory and get an access token to request WorkflowGen's GraphQL API.

File: integration/oidc/authorization-code/README.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+## Prerequisites
+1. dotnet core sdk 2.2
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. .NET Core SDK 2.2

File: integration/oidc/authorization-code/README.md [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+1. Make sure to have a valid WorkflowGen licence and serial number.
+1. Make sure to have access to an Azure subscription.
+1. Make sure to have a WorkflowGen server setup with OIDC authentication with
+Azure AD.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: Make sure to have a WorkflowGen server set up with OIDC authentication with Azure AD.

File: integration/oidc/authorization-code/README.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+    Read the [WorkflowGen for Azure](https://docs.advantys.com/workflowgen-for-azure/) guide to know how!
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: See the [WorkflowGen for Azure](https://docs.advantys.com/workflowgen-for-azure/) guide for information and instructions.

File: integration/oidc/authorization-code/README.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+## Technologies
+
+This example uses aspnet core 2.2 with the OpenIDConnect authentication
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: This example uses ASP.NET Core 2.2 with OpenID Connect authentication

File: integration/oidc/authorization-code/README.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+middleware. For GraphQL requests, it uses the GraphQL.Client library. Pages are
+generated using the Razor engine. Bootstrap and jQuery are available client-side.
+It is based on the dotnet core template named `webapp`.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: It is based on the .NET Core `webapp` template.

File: integration/oidc/authorization-code/README.md [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+
+# Setup
+Before beginning the setup of this example, you should now already have two (2)
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: Before beginning the setup of this example, you should now already have two

File: integration/oidc/authorization-code/README.md [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+app registrations in your Azure AD that represents WorkflowGen's portal and
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: app registrations in your Azure AD that represent the WorkflowGen portal and

File: integration/oidc/authorization-code/README.md [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+WorkflowGen's GraphQL API. If you don't, you must finish the authentication
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: GraphQL API. If you don't, follow the instructions in the authentication

File: integration/oidc/authorization-code/README.md [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+section of the WorkflowGen for Azure guide.
+
+## Create a New App Registration in Azure AD (Azure cli)
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: ## Create a new app registration in Azure AD (Azure CLI)

File: integration/oidc/authorization-code/README.md [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+The following commands create the necessary application registration in
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: The following commands will create the required application registration in

File: integration/oidc/authorization-code/README.md [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+Azure AD using the Azure cli commandline tools.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: Azure AD using the Azure CLI command line tools.

File: integration/oidc/authorization-code/README.md [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+**Note**: You need to be logged in the Azure cli. To do this, run `az login`.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: **Note**: You must be logged into the Azure CLI. To do this, run `az login`.

File: integration/oidc/authorization-code/README.md [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+    --api $webApiApp.appId `
+    --api-permissions "$($webApiApp.oauth2Permissions.id)=Scope"
+
+# Grant the required permissions for the GraphQL API Access
+& az ad app permission grant `
+    --api $webApiApp.appId `
+    --id $webAppClient.appId
+```
+
+You will need the following information for the example configuration:
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: You'll need the following information to configure the example:

File: integration/oidc/authorization-code/README.md [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+
+* The client id. This is the value of `$webAppclient.appId`.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: * The client ID. This is the value of `$webAppclient.appId`.

File: integration/oidc/authorization-code/README.md [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+* The client secret. This is the value of `$password`.
+* The GraphQL API App Id URI. This is the value of `$webApiApp.identifierUris[0]`.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: * The GraphQL API App ID URI. This is the value of `$webApiApp.identifierUris[0]`.

File: integration/oidc/authorization-code/README.md [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+* The Azure AD Directory ID. Run the following command to get it:
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: * The Azure AD Directory ID. To get it, run the following command:

File: integration/oidc/authorization-code/README.md [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+
+    ```powershell
+    az account show | ConvertFrom-Json | ForEach-Object tenantId
+    ```
+
+## Create a New App Registration in Azure AD (Azure Web Portal)
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: ## Create a new app registration in Azure AD (Azure web portal)

File: integration/oidc/authorization-code/README.md [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+To represent this example application that wants to access WorkflowGen's
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: To represent this example application that wants to access the WorkflowGen

File: integration/oidc/authorization-code/README.md [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+GraphQL API, we need to register it to Azure AD and give it access to the API.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: GraphQL API, you need to register it in Azure AD and give it access to the API.

File: integration/oidc/authorization-code/README.md [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+This way, when requesting the GraphQL API resource in the OIDC protocol, you
+will get an access token that is a valid JWT and that can be transmitted to
+the API for authorization.
+
+Keep in mind that the user that will authenticate with this application must
+exists in WorkflowGen.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: exist in WorkflowGen.

File: integration/oidc/authorization-code/README.md [hunk 23]
```diff
@@ -0,0 +1,197 @@
...
+
+1. In Azure AD, go to the `App Registrations` section.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. In Azure AD, go to the **App Registrations** section and click the **New application registration** button.

File: integration/oidc/authorization-code/README.md [hunk 24]
```diff
@@ -0,0 +1,197 @@
...
+1. Click on the `New application registration` button.
+1. Enter the following information:
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: remove

File: integration/oidc/authorization-code/README.md [hunk 25]
```diff
@@ -0,0 +1,197 @@
...
+GraphQL API, we need to register it to Azure AD and give it access to the API.
+This way, when requesting the GraphQL API resource in the OIDC protocol, you
+will get an access token that is a valid JWT and that can be transmitted to
+the API for authorization.
+
+Keep in mind that the user that will authenticate with this application must
+exists in WorkflowGen.
+
+1. In Azure AD, go to the `App Registrations` section.
+1. Click on the `New application registration` button.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: remove

File: integration/oidc/authorization-code/README.md [hunk 26]
```diff
@@ -0,0 +1,197 @@
...
+1. Enter the following information:
+
+    Name: `My aspnet core Web App`
+    Application Type: `Web app / API`
+    Sign-on URL: `http://localhost:8080`
+
+1. When you are ready, click on the `Create` button.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. When you're done, click the `Create` button.

File: integration/oidc/authorization-code/README.md [hunk 27]
```diff
@@ -0,0 +1,197 @@
...
+
+You now have registered this example application. You now need to finish
+configuring it in order to get the auhtentication working properly.
+
+1. Go to the `My aspnet core Web App` application registration page in Azure AD.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Go to the **My aspnet core Web App** application registration page in Azure AD.

^^^^ Replace **My aspnet core Web App** with exactly what appears on the page.

File: integration/oidc/authorization-code/README.md [hunk 28]
```diff
@@ -0,0 +1,197 @@
...
+1. Copy the Application ID and save it for later use. This is the `Client Id`
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Copy the Application ID, which you'll need later. This is the application's Client ID.

File: integration/oidc/authorization-code/README.md [hunk 29]
```diff
@@ -0,0 +1,197 @@
...
+of the application.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: remove

File: integration/oidc/authorization-code/README.md [hunk 30]
```diff
@@ -0,0 +1,197 @@
...
+1. Click on the `Settings` button.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Click the **Settings** button.

File: integration/oidc/authorization-code/README.md [hunk 31]
```diff
@@ -0,0 +1,197 @@
...
+1. Go to the `Reply URLs` section.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. In the **Reply URLs** section, add `http://localhost:8080/signin` to the list, then click **Save**.

File: integration/oidc/authorization-code/README.md [hunk 32]
```diff
@@ -0,0 +1,197 @@
...
+1. Add `http://localhost:8080/signin` to the list and click on the `Save`
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: remove

File: integration/oidc/authorization-code/README.md [hunk 33]
```diff
@@ -0,0 +1,197 @@
...
+button.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: remove

File: integration/oidc/authorization-code/README.md [hunk 34]
```diff
@@ -0,0 +1,197 @@
...
+1. Return to the sections list and go to the `Keys` section.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Return to the sections list and go to the **Keys** section.

File: integration/oidc/authorization-code/README.md [hunk 35]
```diff
@@ -0,0 +1,197 @@
...
+1. In the `Passwords` sub-section, add a line with the following information:
+
+    Description: `secret`
+    Expires: `Never expires`
+    Value: A random Guid (see the following note)
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: Value: A random GUID (see the following note)

File: integration/oidc/authorization-code/README.md [hunk 36]
```diff
@@ -0,0 +1,197 @@
...
+
+    **About Generating a Guid**
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: **How to generate a GUID**

File: integration/oidc/authorization-code/README.md [hunk 37]
```diff
@@ -0,0 +1,197 @@
...
+
+    Open a PowerShell terminal and enter the following command:
+    ```powershell
+    [guid]::NewGuid().ToString()
+    ```
+    Copy the result and enter in the value section of the form.
+
+1. Click on the `Save` button.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Click **Save**.

File: integration/oidc/authorization-code/README.md [hunk 38]
```diff
@@ -0,0 +1,197 @@
...
+1. Copy the value that has been generated after the save. It is the `Client Secret`
+of this application.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: remove

File: integration/oidc/authorization-code/README.md [hunk 39]
```diff
@@ -0,0 +1,197 @@
...
+1. Return to the sections list and go to the `Required permissions` section.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Return to the sections list and go to the **Required permissions** section.

File: integration/oidc/authorization-code/README.md [hunk 40]
```diff
@@ -0,0 +1,197 @@
...
+1. Click on the `Add` button.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Click **Add** then **Select an API**.

File: integration/oidc/authorization-code/README.md [hunk 41]
```diff
@@ -0,0 +1,197 @@
...
+1. Click on the `Select an API` button.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: remove

File: integration/oidc/authorization-code/README.md [hunk 42]
```diff
@@ -0,0 +1,197 @@
...
+1. Search for the WorkflowGen's GraphQL API application that you've configured
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Search for the WorkflowGen GraphQL API application that you configured

File: integration/oidc/authorization-code/README.md [hunk 43]
```diff
@@ -0,0 +1,197 @@
...
+prior to this guide, click on it and click on the `Select` button.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: remove

File: integration/oidc/authorization-code/README.md [hunk 44]
```diff
@@ -0,0 +1,197 @@
...
+1. In the `Select permissions` section, check all checkboxes and click on the
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. In the **Select permissions** section, check all checkboxes, then click the **Select** button.

File: integration/oidc/authorization-code/README.md [hunk 45]
```diff
@@ -0,0 +1,197 @@
...
+`Select` button.
+1. Click on the `Done` button.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Click the **Done** button.

File: integration/oidc/authorization-code/README.md [hunk 46]
```diff
@@ -0,0 +1,197 @@
...
+
+You should now see WorkflowGen's GraphQL API in the list of required permissions
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: You should now see the WorkflowGen GraphQL API in the list of required permissions

File: integration/oidc/authorization-code/README.md [hunk 47]
```diff
@@ -0,0 +1,197 @@
...
+for this example application.
+
+You should now have four (4) pieces of information in your possession at this
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: You should now have four pieces of information:

File: integration/oidc/authorization-code/README.md [hunk 48]
```diff
@@ -0,0 +1,197 @@
...
+point:
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: remove

File: integration/oidc/authorization-code/README.md [hunk 49]
```diff
@@ -0,0 +1,197 @@
...
+
+1. This application registration's Application ID or `Client Id`
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. The application registration's Application ID or Client ID

File: integration/oidc/authorization-code/README.md [hunk 50]
```diff
@@ -0,0 +1,197 @@
...
+1. This application registration's Password Key or `Client Secret`
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. The application registration's Password Key or Client Secret

File: integration/oidc/authorization-code/README.md [hunk 51]
```diff
@@ -0,0 +1,197 @@
...
+1. The App ID URI of WorkflowGen's GraphQL API. It should look like this:
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. The WorkflowGen GraphQL API's App ID URI, which should look like this:

File: integration/oidc/authorization-code/README.md [hunk 52]
```diff
@@ -0,0 +1,197 @@
...
+
+    `http://localhost:8888/wfgen/graphql`
+
+1. Your Azure AD Directory ID which can be found in the properties section of your
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Your Azure AD Directory ID, which can be found in the **Properties** section of your Azure AD page.

File: integration/oidc/authorization-code/README.md [hunk 53]
```diff
@@ -0,0 +1,197 @@
...
+Azure AD page.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: remove

File: integration/oidc/authorization-code/README.md [hunk 54]
```diff
@@ -0,0 +1,197 @@
...
+
+## Configure the Example Application
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: ## Configure the example application

File: integration/oidc/authorization-code/README.md [hunk 55]
```diff
@@ -0,0 +1,197 @@
...
+The configuration of the example is straightforward and takes a lot less steps
+than configuring it in Azure AD.
+
+### Prerequisites
+1. Make sure to have installed the dotnet core sdk version 2.2. The `dotnet`
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Make sure to have installed .NET Core SDK version 2.2. The `dotnet`

File: integration/oidc/authorization-code/README.md [hunk 56]
```diff
@@ -0,0 +1,197 @@
...
+commandline tool should be available in a terminal.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: command line tool should be available in a terminal.

File: integration/oidc/authorization-code/README.md [hunk 57]
```diff
@@ -0,0 +1,197 @@
...
+1. Make sure that the user that you're using is a valid WorkflowGen user.
+
+Optionally, you can install Visual Studio Code to run and debug this example.
+There is existing vscode settings in the repository to make it easier to get
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: There are existing VSCode settings in the repository to make it easier to get

File: integration/oidc/authorization-code/README.md [hunk 58]
```diff
@@ -0,0 +1,197 @@
...
+started.
+
+### Get Started
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: ### Get started

^^^ this implies there's another section after this one. What are they actually doing here?

File: integration/oidc/authorization-code/README.md [hunk 59]
```diff
@@ -0,0 +1,197 @@
...
+1. Clone this application somewhere on your computer.
+
+    ```powershell
+    git clone https://github.com/advantys/workflowgen-templates.git
+    Set-Location .\workflowgen-templates\integration\oidc\authorization-code
+    ```
+
+1. Open the project folder with vscode.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Open the project folder with VSCode.

File: integration/oidc/authorization-code/README.md [hunk 60]
```diff
@@ -0,0 +1,197 @@
...
+1. Duplicate the file `appsettings.Development.template.json` and rename it to
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Duplicate the `appsettings.Development.template.json` file and rename it as

File: integration/oidc/authorization-code/README.md [hunk 61]
```diff
@@ -0,0 +1,197 @@
...
+`appsettings.Development.json`.
+1. Duplicate the file `appsettings.template.json` and rename it to
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Duplicate the `appsettings.template.json` file and rename it as

File: integration/oidc/authorization-code/README.md [hunk 62]
```diff
@@ -0,0 +1,197 @@
...
+`appsettings.json`.
+1. In the `appsettings.json` file, replace the placeholders surrounded by `[]`
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. In the `appsettings.json` file, replace the placeholders inside brackets (`[]`)

File: integration/oidc/authorization-code/README.md [hunk 63]
```diff
@@ -0,0 +1,197 @@
...
+inclusively with the values that you have gattered in the last section.
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: with the values that you got in the previous section.

File: integration/oidc/authorization-code/README.md [hunk 64]
```diff
@@ -0,0 +1,197 @@
...
+
+    * `[your azure ad tenant id]`: Your Azure AD Directory ID
+    * `[your application's client id (Application Id)]`: The Client Id
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: * `[your application's client id (Application Id)]`: The Client ID

File: integration/oidc/authorization-code/README.md [hunk 65]
```diff
@@ -0,0 +1,197 @@
...
+    * `[your application's client secret added to your azure ad app]`: The Client Secret
+    * `[workflowgen's url]`: The part of the App ID URI which represents WorkflowGen's
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: * `[workflowgen's url]`: The part of the App ID URI that represents WorkflowGen's

^^^ if possible change placeholder to `[workflowgen url]`

File: integration/oidc/authorization-code/README.md [hunk 66]
```diff
@@ -0,0 +1,197 @@
...
+    domain and base path e.g. `localhost:8888/wfgen` in the example App ID URI
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: domain and base path (e.g. `localhost:8888/wfgen`) in the example App ID URI

File: integration/oidc/authorization-code/README.md [hunk 67]
```diff
@@ -0,0 +1,197 @@
...
+    above.
+
+1. Start debugging in vscode or run:
```
Comment (pnicoll) [inline integration/oidc/authorization-code/README.md]: 1. Start debugging in VSCode, or run the following commands:

**Pull request merged**